torch/distributed/_sharded_tensor/api.py

```diff
@@ -309,6 +307,7 @@ def gather(
                     tensor = shard.tensor
 
                     out_narrow_view = out
+                    assert out_narrow_view is not None
```

Comment:
Doesn't `_validate_output_tensor_for_gather` validate this? Why do we need another assert here?

Reply:
this is added purely for mypy linter, it seems like mypy couldn't understand `_validate_output_tensor_for_gather` checks, so have to do this assert here.

Reply:
Usually I add a mypy ignore for something like this, but its up to you.